### ISSUE TITLE
Gradle not honoring Maven exclusions defined in pom.xml

### ISSUE THREAD

**Issue by vodasams57** [gradle/gradle#22370]
### Expected Behavior
Gradle does not exclude libraries defined in Maven dependencies (pom.xml).

### Current Behavior
In contrast to Maven, Gradle adds excluded dependencies to the dependency tree.

### Context
We use Apache James mailserver and switched from Maven to Gradle build system. We ran into a dependency conflict, because `org.apache.james:apache-mailet:2.4` was also in the runtime classpath in addition to `org.apache.james:apache-mailet-base:3.7.1`. Having `org.apache.james:apache-mailet:2.4` in the runtime classpath caused a runtime error. After investigating we noticed `org.apache.james:apache-mailet:2.4` was excluded in the corresponding [pom.xml](https://search.maven.org/artifact/org.apache.james/james-server-mailet-dkim/3.7.1/jar) of `org.apache.james:james-server-mailet-dkim`. 

After browsing Gradle issues to gain clarity, we followed the discussion in #12214 and [spring-boot#21350](https://github.com/spring-projects/spring-boot/issues/21350). We accepted the fact that Maven and Gradle use different dependency resolution strategies. However, we filed this bug report to ask for clarity how to tackle such issues:
1. It is unclear why the exclusion defined in the [pom.xml](https://search.maven.org/artifact/org.apache.james/james-server-mailet-dkim/3.7.1/jar) lands in the dependency tree when building with Gradle. We distilled the example to a simple build.gradle and pom.xml (see "Steps to Reproduce"). Can someone explain this behavior? My vague assumption is that the cause is a circular dependency of `org.apache.james:apache-mailet` (see "Steps to Reproduce" `dependencyInsight`) or the `<parent />` tag of [pom.xml](https://search.maven.org/artifact/org.apache.james/james-server-mailet-dkim/3.7.1/jar) that references to an additional pom.xml.
2. Do you consider this expected behavior?
3. This has been discussed before to some extend in [#12214#issuecomment-585888556](https://github.com/gradle/gradle/issues/12214#issuecomment-585888556). Does this also apply to this issue?
4. What is the best strategy to handle such dependencies? An alternative would be to switch back to Maven, so exclusions defined by maintainers of Apache James are acknowledged.

### Steps to Reproduce 

simple build.gradle
```
plugins {
    id 'java'
}

repositories {
    mavenCentral()
}

dependencies {
    implementation "org.apache.james:james-server-mailet-dkim:3.7.1"
}

group = 'com.example'
description = 'gradletest'
```

Maven counterpart (pom.xml)
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>com.example</groupId>
  <artifactId>gradletest</artifactId>
  <version>1.0-SNAPSHOT</version>

  <dependencies>
    <dependency>
      <groupId>org.apache.james</groupId>
      <artifactId>james-server-mailet-dkim</artifactId>
      <version>3.7.1</version>
      <scope>runtime</scope>
    </dependency>
  </dependencies>

</project>
```

Dependency tree generated by maven that **does not include** `org.apache.james:apache-mailet:2.4` and other exclusions such as `org.apache.james:apache-mailet-base:1.1`.
```
> mvn dependency:tree

[INFO] com.example:gradletest:jar:1.0-SNAPSHOT
[INFO] \- org.apache.james:james-server-mailet-dkim:jar:3.7.1:runtime
[INFO]    +- org.apache.james:apache-mailet-api:jar:3.7.1:runtime
...
[INFO]    +- org.apache.james.jdkim:apache-jdkim-library:jar:0.2:runtime
[INFO]    |  +- org.apache.geronimo.specs:geronimo-activation_1.1_spec:jar:1.0.2:runtime
[INFO]    |  \- dnsjava:dnsjava:jar:2.1.1:runtime
[INFO]    |     \- junit:junit:jar:3.8.2:runtime
...
```

Dependency tree generated by Gradle that **includes** `org.apache.james:apache-mailet:2.4` and `org.apache.james:apache-mailet-base:1.1`
```
> gradle dependencies --configuration runtimeClasspath

runtimeClasspath - Runtime classpath of source set 'main'.
\--- org.apache.james:james-server-mailet-dkim:3.7.1
     +--- org.apache.james:apache-mailet-api:3.7.1
...
     +--- org.apache.james.jdkim:apache-jdkim-library:0.2
     |    +--- org.apache.geronimo.specs:geronimo-activation_1.1_spec:1.0.2
     |    +--- org.apache.james:apache-mailet-base:1.1 -> 3.7.1 (*)
     |    +--- org.apache.james:apache-mailet:2.4
     |    \--- dnsjava:dnsjava:2.1.1
     |         \--- junit:junit:3.8.2
...
```

Dependency insight generated by Gradle

```
> gradle -q dependencyInsight --dependency org.apache.james:apache-mailet --configuration runtimeClasspath
org.apache.james:apache-mailet:2.4
   variant "runtime" [
      org.gradle.status              = release (not requested)
      org.gradle.usage               = java-runtime
      org.gradle.libraryelements     = jar
      org.gradle.category            = library

      Requested attributes not found in the selected variant:
         org.gradle.dependency.bundling = external
         org.gradle.jvm.environment     = standard-jvm
         org.gradle.jvm.version         = 8
   ]

org.apache.james:apache-mailet:2.4
\--- org.apache.james.jdkim:apache-jdkim-library:0.2
     \--- org.apache.james:james-server-mailet-dkim:3.7.1
          \--- runtimeClasspath

org.apache.james:apache-mailet-api:3.7.1
   variant "runtime" [
      org.gradle.status              = release (not requested)
      org.gradle.usage               = java-runtime
      org.gradle.libraryelements     = jar
      org.gradle.category            = library

      Requested attributes not found in the selected variant:
         org.gradle.dependency.bundling = external
         org.gradle.jvm.environment     = standard-jvm
         org.gradle.jvm.version         = 8
   ]

org.apache.james:apache-mailet-api:3.7.1
+--- org.apache.james:apache-mailet-base:3.7.1
|    +--- org.apache.james:james-server-mailet-dkim:3.7.1
|    |    \--- runtimeClasspath
|    \--- org.apache.james.jdkim:apache-jdkim-library:0.2 (requested org.apache.james:apache-mailet-base:1.1)
|         \--- org.apache.james:james-server-mailet-dkim:3.7.1 (*)
\--- org.apache.james:james-server-mailet-dkim:3.7.1 (*)

org.apache.james:apache-mailet-base:3.7.1
   variant "runtime" [
      org.gradle.status              = release (not requested)
      org.gradle.usage               = java-runtime
      org.gradle.libraryelements     = jar
      org.gradle.category            = library

      Requested attributes not found in the selected variant:
         org.gradle.dependency.bundling = external
         org.gradle.jvm.environment     = standard-jvm
         org.gradle.jvm.version         = 8
   ]
   Selection reasons:
      - By conflict resolution : between versions 3.7.1 and 1.1

org.apache.james:apache-mailet-base:3.7.1
\--- org.apache.james:james-server-mailet-dkim:3.7.1
     \--- runtimeClasspath

org.apache.james:apache-mailet-base:1.1 -> 3.7.1
\--- org.apache.james.jdkim:apache-jdkim-library:0.2
     \--- org.apache.james:james-server-mailet-dkim:3.7.1
          \--- runtimeClasspath

(*) - dependencies omitted (listed previously)
```

### Your Environment

```
> gradle --version            

------------------------------------------------------------
Gradle 7.1
------------------------------------------------------------

Build time:   2021-06-14 14:47:26 UTC
Revision:     989ccc9952b140ee6ab88870e8a12f1b2998369e

Kotlin:       1.4.31
Groovy:       3.0.7
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          1.8.0_312 (BellSoft 25.312-b07)
OS:           Mac OS X 12.5.1 aarch64
```

**Comment by ljacomet** [comment]
This is actually a bug in Gradle where we do not handle properties inside Maven POM `exclusions`.

And since you have the following in the [POM of `james-server-mailet-dkim`](https://repo.maven.apache.org/maven2/org/apache/james/james-server-mailet-dkim/3.7.1/james-server-mailet-dkim-3.7.1.pom):
```
<dependency>
    <groupId>org.apache.james.jdkim</groupId>
    <artifactId>apache-jdkim-library</artifactId>
    <version>${jdkim.version}</version>
    <exclusions>
        <exclusion>
            <groupId>${james.groupId}</groupId>
            <artifactId>apache-mailet</artifactId>
        </exclusion>
        ...
    </exclusions>
</dependency>
```

So Gradle needs to handle property expansions in `exclusions` declarations.

The simplest workaround is to [declare that exclusion](https://docs.gradle.org/current/userguide/dependency_downgrade_and_exclude.html#sec:excluding-transitive-deps) when adding the dependency on `james-server-mailet-dkim`.

**Comment by vodasams57** [comment]
Thanks for confirming the bug report.

We already declared the exclusion in the build.gradle, thanks for the hint.

After having the runtime error we broke it down to `james-server-mailet-dkim` problem. This allowed me to file this bug report with a simple reproducible example. Since we have a lot of Apache James dependencies there could be more potential issues. Now we need to compare the resolved dependency trees between Maven and Gradle and look for missing exclusions.

Is there a easier way than to compare the dependency tree output (i.e. `gradle dependencies --configuration runtimeClasspath` vs. `mvn dependency:tree`)?
